pkg/scheduler/algorithm/priorities/resource_bin_packing.go

```diff
@@ -0,0 +1,240 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	priorityutil "k8s.io/kubernetes/pkg/scheduler/algorithm/priorities/util"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+// ResourceBinPacking contains information to calculate bin packing priority.
+type ResourceBinPacking struct {
+	Resources []Resources
+}
+
+//Resources contains resource name and weight
+type Resources struct {
+	Resource v1.ResourceName
+	Weight   int
+}
+
+// NewResourceBinPacking creates a ResourceBinPackingPriorityMap.
+func NewResourceBinPacking(resources []Resources) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	for i, elm := range resources {
+		if elm.Weight == 0 {
+			resources[i].Weight = 1
+		}
+	}
+	resourceBinPackingPrioritizer := &ResourceBinPacking{
+		Resources: resources,
+	}
+	return resourceBinPackingPrioritizer.ResourceBinPackingPriorityMap, nil
+}
+
+// ResourceBinPackingPriorityDefault creates a resourceBinPacking based prioity
+func ResourceBinPackingPriorityDefault() *ResourceBinPacking {
+	defaultResourceBinPackingPrioritizer := &ResourceBinPacking{
+		Resources: []Resources{},
+	}
+	return defaultResourceBinPackingPrioritizer
+}
+
+// ResourceBinPackingPriorityMap is a priority function that favors nodes that have higher utlization of scare resource.
+// It will detect whether the requested scarce resource is present on a node, and then calculate a score ranging from 0 to 10
+// based total utlization (best fit)
+// - If none of the scare resource are requested, this node will be given the lowest priority.
+// - If the scarce resource is requested, the larger the resource utlization ratio, the higher the node's priority.
+func (r *ResourceBinPacking) ResourceBinPackingPriorityMap(pod *v1.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error) {
+	var score int
+	node := nodeInfo.Node()
+	if len(r.Resources) == 0 {
+		return schedulerapi.HostPriority{}, fmt.Errorf("resource not defined")
+	}
+	if node == nil {
+		return schedulerapi.HostPriority{}, fmt.Errorf("node not found")
+	}
+	nodeScore := float64(0)
+	weightSum := 0
+
+	for _, elm := range r.Resources {
+		nodeScore += calculateScareResourceScore(nodeInfo, pod, elm.Resource) * float64(elm.Weight)
+		weightSum += elm.Weight
```

Comment:
If pod does not request the resource, do we need to account it into weightSum?

Reply:
yes, that's a good catch. I'll update that to consider only when the response is non-zero